src/Umbraco.Web.Common/Media/ImageSharpImageUrlTokenGenerator.cs

```diff
@@ -0,0 +1,95 @@
+    /// Exposes a method that generates an image URL token for request authentication.
+    /// </summary>
+    /// <seealso cref="IImageUrlTokenGenerator" />
+    public class ImageSharpImageUrlTokenGenerator : IImageUrlTokenGenerator
+    {
+        private readonly byte[]? _hmacSecretKey;
+        private readonly Lazy<IList<string>?> _knownCommands;
+        private readonly CaseHandlingUriBuilder.CaseHandling _caseHandling;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ImageSharpImageUrlTokenGenerator" /> class.
+        /// </summary>
+        /// <param name="imagingSettings">The Umbraco imaging settings.</param>
+        /// <param name="imageWebProcessors">"The image web processors to retrieve the known commands, used to strip out unsupported commands from the generated image URL token.</param>
+        public ImageSharpImageUrlTokenGenerator(IOptions<ImagingSettings> imagingSettings, Lazy<IEnumerable<IImageWebProcessor>> imageWebProcessors)
+            : this(imagingSettings.Value.HMACSecretKey, imageWebProcessors, CaseHandlingUriBuilder.CaseHandling.LowerInvariant)
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ImageSharpImageUrlTokenGenerator" /> class.
+        /// </summary>
+        /// <param name="hmacSecretKey">The HMAC security key.</param>
+        /// <param name="knownCommands">"The known commands, used to strip out unsupported commands from the generated image URL token.</param>
+        /// <remarks>
+        /// This constructor is only used for testing.
+        /// </remarks>
+        internal ImageSharpImageUrlTokenGenerator(byte[]? hmacSecretKey, IList<string>? knownCommands = null)
+            : this(hmacSecretKey, null, CaseHandlingUriBuilder.CaseHandling.LowerInvariant)
+        {
+            _knownCommands = new Lazy<IList<string>?>(knownCommands);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ImageSharpImageUrlTokenGenerator" /> class.
+        /// </summary>
+        /// <param name="hmacSecretKey">The HMAC security key.</param>
+        /// <param name="imageWebProcessors">"The image web processors to retrieve the known commands, used to strip out unsupported commands from the generated image URL token.</param>
+        /// <param name="caseHandling">Determines case handling for the result.</param>
+        protected ImageSharpImageUrlTokenGenerator(byte[]? hmacSecretKey, Lazy<IEnumerable<IImageWebProcessor>>? imageWebProcessors, CaseHandlingUriBuilder.CaseHandling caseHandling)
+        {
+            _hmacSecretKey = hmacSecretKey;
+            _knownCommands = new Lazy<IList<string>?>(() => imageWebProcessors?.Value.SelectMany(x => x.Commands).Distinct().ToList());
+            _caseHandling = caseHandling;
+        }
+
+        /// <inheritdoc />
+        public string? GetImageUrlToken(string imageUrl, IEnumerable<KeyValuePair<string, string?>> commands)
```

Comment:
This all seems very complicated. It should be as simple as using the relative URL as a string and pass it to the new Encode method on the url builder. I added that for convenience in the last release but the operations of that method would have been how I would have handled it without it.

Reply:
Thanks for letting me know about the new `Encode()` method, although it seems like that doesn't lowercase the query string/commands for relative URLs when using `CaseHandling.None` (and it doesn't remove unknown commands from the URL for you).

And although it might seem a bit complicated to abstract image URL token generation, this decouples it from ImageSharp (e.g. in case you want to use a cloud service), ensures both generation and validation uses the same logic and enables users to overwrite the default implementation.

Reply:
ImageSharp validates the HMAC token using using only known commands (as unknown commands are stripped before being passed to the `OnComputeHMACAsync` callback), so you also need to ensure you don't include unknown commands while generating the token.

Umbraco provides a way to add additional commands using `furtherOptions` and if you remove processors, those commands will become unknown and stripped, so Umbraco can't maintain control over this...

Reply:
Would something like this make life easier for you?

I'm tinkering with a singleton injected type that would unify everything so that you can simply pass a relative or absolute string/URI and it will generate a token based upon the options.

<img alt="image" src="https://user-images.githubusercontent.com/385879/177348613-60d0f493-a990-47b3-a0d2-5cbabbf576dd.png">

Reply:
@JimBobSquarePants I just checked out your `js/auth-utils` branch on ImageSharp.Web and the `ImageSharpRequestAuthorizationUtilities` would certainly help make things easier! 🙌🏻

I know this is still a work-in-progress, but already wanted to give my early feedback:
- Update the constructor:
  - Ensure the `requestParser` and `serviceProvider` fields are set;
  - Possibly skip initializing the known commands when no secret is set;
  - Nit-picking: change the order of parameters and field setters to `options`, `serviceProvider`, `requestParser` and `processors` (the order in which they are used in `ComputeHMACAsync`)
- Add sync/async overloads accepting `HostString host, PathString path, QueryString queryString` and `HostString host, PathString path, IQueryCollection query`;
- Possibly already register the `ImageSharpRequestAuthorizationUtilities`, so it can be retrieved from the service container;
- Nit-picking: use `path` instead of `pathBase` in `CaseHandlingUriBuilder.Encode()`.

Reply:
For context: the above mentioned branch is now converted into a PR: https://github.com/SixLabors/ImageSharp.Web/pull/281.

Reply:
I've converted this PR back to draft, awaiting a new release of ImageSharp.Web that includes the beforementioned PR with authorization utilities...